File: sdks/swift/Sources/Skopli/Skopli.swift
import Foundation

// The idiomatic Swift facade over the C ABI. Pure Swift: Codable models in/out,
// throws for errors, camelCase throughout. Marshalling goes through the 15-fn
// JSON surface via the Ffi bridge; nothing here touches CSkopli directly.
//
// Async conveniences run the sync primitive off the calling thread via
// withCheckedThrowingContinuation on a background queue.

/// A single shared encoder/decoder pair. JSONEncoder omits nil optionals by
/// default, matching the wire's absent-vs-null discipline.
private let encoder = JSONEncoder()
private let decoder = JSONDecoder()

private let workQueue = DispatchQueue(
    label: "com.skopli.work",
    qos: .userInitiated,
    attributes: .concurrent
)

private func encode<T: Encodable>(_ value: T) throws -> Data {
    do {
        return try encoder.encode(value)
    } catch {
        throw SkopliError.invalidArgument("encoding options: \(error)")
    }
}

private func decode<T: Decodable>(_ type: T.Type, from data: Data) throws -> T {
    do {
        return try decoder.decode(type, from: data)
    } catch {
        throw SkopliError.internalError("decoding native response: \(error)")
    }
}

/// The top-level Skopli operations. Static methods mirror the common facade
/// contract: detect/read/rollup/cost/pricing.
public enum Skopli {
    // MARK: - Meta

    /// The native ABI version integer (bumped on a breaking ABI change).
    public static var abiVersion: UInt32 { Ffi.abiVersion() }

    /// The conformance `schema_version` the native build emits.
    public static var schemaVersion: UInt32 { Ffi.schemaVersion() }

    /// The native library semver string.
    public static var version: String { Ffi.version() }

    /// The default on-disk pricing cache directory (platform-specific).
    public static func defaultCacheDir() throws -> String {
        try Ffi.defaultCacheDir()
    }

    // MARK: - Pipeline (sync)

    /// Report which registered harnesses have readable data reachable from
    /// `options`. `nil` options uses process defaults.
    public static func detectHarnesses(options: Options? = nil) throws -> Detection {
        let opts = try options.map(encode)
        let out = try Ffi.detectHarnesses(opts)
        return try decode(Detection.self, from: out)
    }

    /// Read usage across the selected harnesses, returning the
    /// `{events, diagnostics, skipped}` envelope. Malformed input never throws;
    /// it is reported in `diagnostics`/`skipped`.
    public static func readUsage(options: ReadUsageOptions = ReadUsageOptions()) throws
        -> ReadUsageResult
    {
        let out = try Ffi.readUsage(try encode(options))
        return try decode(ReadUsageResult.self, from: out)
    }

    /// Aggregate events by the option dimension. `blockMs` sets the
    /// billing-block width for `.block` (defaults to five hours) and is ignored
    /// for other dimensions.
    public static func rollup(
        events: [UsageEvent], by: RollupBy, tz: String? = nil, blockMs: Int64? = nil
    ) throws -> [Rollup] {
        let eventsData = try encode(events)
        let optsData = try encode(RollupOptions(by: by, tz: tz, blockMs: blockMs))
        let out = try Ffi.rollup(events: eventsData, opts: optsData)
        return try decode([Rollup].self, from: out)
    }

    /// Compute the USD cost of a token bundle under a flat or tiered price.
    public static func costUSD(tokens: TokenCounts, price: ModelPrice) throws -> Double {
        let tokensData = try encode(tokens)
        let priceData = try encode(price)
        return try Ffi.costUsd(tokens: tokensData, price: priceData)
    }

    /// Build a `Pricing` handle from the given options (resolves any
    /// facade-side `PricingSource`s first; see `PricingOptions`).
    public static func createPricing(options: PricingOptions = PricingOptions()) throws -> Pricing {
        try Pricing(options: options)
    }

    // MARK: - Pipeline (async conveniences)

    /// Async `readUsage`, running the sync primitive off-thread.
    public static func readUsage(options: ReadUsageOptions = ReadUsageOptions()) async throws
        -> ReadUsageResult
    {
        try await runOffThread { try readUsage(options: options) }
    }

    /// Async `createPricing`, running the sync primitive off-thread.
    public static func createPricing(options: PricingOptions = PricingOptions()) async throws
        -> Pricing
    {
        try await runOffThread { try createPricing(options: options) }
    }
}

/// Run a throwing sync closure on the background queue, bridging to async via a
/// checked continuation: the sync core runs off-thread.
func runOffThread<T: Sendable>(_ work: @escaping @Sendable () throws -> T) async throws -> T {
    try await withCheckedThrowingContinuation { continuation in
        workQueue.async {
            continuation.resume(with: Result { try work() })
        }
    }
}

File: sdks/java/src/main/java/com/skopli/Skopli.java
package com.skopli;

import com.skopli.internal.Ffi;
import com.skopli.internal.Wire;
import com.skopli.json.Json;
import java.util.ArrayList;
import java.util.List;

/**
 * The skopli SDK entry point: static operations over the C ABI (ridden via
 * Java FFM). Sync and thread-safe; wrap in {@link java.util.concurrent.CompletableFuture}
 * for async. All data crosses as JSON marshalled to/from the idiomatic records
 * in this package; a non-zero native status surfaces as an unchecked
 * {@link SkopliException}.
 */
public final class Skopli {

    private Skopli() {}

    /** Force the native library to load eagerly (otherwise it loads on first
     * call). Useful to surface an {@link UnsatisfiedLinkError} at startup. */
    public static void load() {
        Ffi.init();
    }

    /** The ABI version integer (bumped on a breaking ABI change). */
    public static int abiVersion() {
        return Ffi.abiVersion();
    }

    /** The conformance schema version this build emits. */
    public static int schemaVersion() {
        return Ffi.schemaVersion();
    }

    /** The library semver string. */
    public static String version() {
        return Ffi.version();
    }

    /** The default on-disk cache directory (platform-specific). */
    public static String defaultCacheDir() {
        return Ffi.defaultCacheDir();
    }

    /** Detect which registered harnesses have readable data reachable from the
     * context. */
    public static Detection detectHarnesses(DetectOptions options) {
        String optsJson = Json.write(options.toWire());
        String out = Ffi.optsToJson(com.skopli.ffi.Skopli::ag_detect_harnesses, optsJson);
        return Detection.fromWire(Wire.asObject(Json.parse(out)));
    }

    /** Read usage across the selected harnesses, returning the
     * {events, diagnostics, skipped} envelope. */
    public static ReadUsageResult readUsage(ReadUsageOptions options) {
        String optsJson = Json.write(options.toWire());
        String out = Ffi.optsToJson(com.skopli.ffi.Skopli::ag_read_usage, optsJson);
        return ReadUsageResult.fromWire(Wire.asObject(Json.parse(out)));
    }

    /** Roll up an event batch by the option-named dimension. */
    public static List<Rollup> rollup(List<UsageEvent> events, RollupOptions options) {
        String eventsJson = eventsToJson(events);
        String optsJson = Json.write(options.toWire());
        String out = Ffi.rollup(eventsJson, optsJson);
        List<Rollup> result = new ArrayList<>();
        for (Object r : Wire.asArray(Json.parse(out))) {
            result.add(Rollup.fromWire(Wire.asObject(r)));
        }
        return List.copyOf(result);
    }

    /** Compute the USD cost of a token bundle under a flat/tiered price. */
    public static double costUsd(TokenCounts tokens, ModelPrice price) {
        String tokensJson = Json.write(tokens.toWire());
        String priceJson = Json.write(price.toWire());
        return Ffi.costUsd(tokensJson, priceJson);
    }

    /** Construct a {@link Pricing} handle. Any {@link PricingSource}s are
     * resolved facade-side and their catalogs merged before the FFI call. Use
     * try-with-resources ({@link Pricing} is {@link AutoCloseable}). */
    public static Pricing createPricing(PricingOptions options) {
        return Pricing.create(options);
    }

    static String eventsToJson(List<UsageEvent> events) {
        List<Object> arr = new ArrayList<>(events.size());
        for (UsageEvent e : events) {
            arr.add(e.toWire());
        }
        return Json.write(arr);
    }
}
